src/user/tag.ts

```diff
@@ -0,0 +1,182 @@
+      return existingTag
+    }
+
+    const newTag = new (this as any)(id) as Tag
+    this.pool.set(id, newTag)
+
+    return newTag
+  }
+
+  /**
+   * Get all tags of this account
+   *
+   * > Tips:
+   * This function is depending on the Puppet Implementation, see [puppet-compatible-table](https://github.com/Chatie/wechaty/wiki/Puppet#3-puppet-compatible-table)
+   *
+   * @static
+   * @returns {Promise<Tag[]>}
+   * @example
+   * const bot = new Wechaty()
+   * await bot.start()
+   * await bot.Tag.findAll()
+   */
+  public static async findAll<T extends typeof Tag> (
+    this: T,
+    target: typeof Contact | typeof Favorite,
+  ): Promise<Array<T['prototype']>> {
+    log.verbose('Tag', 'findAll(%s)', target)
+
+    const tagList: Tag[] = []
+
+    if (target === Contact || target === this.wechaty.Contact) {
+      const contactList = await this.wechaty.Contact.findAll()
+      for (const contact of contactList) {
+        const contactTagList = await contact.tags()
+        contactTagList.forEach(tag => {
+          if (!tagList.includes(tag)) {
+              tagList.push(tag)
+            }
+          })
+      }
+    // } else if (target === Favorite || target === this.wechaty.Favorite)
+    } else {
+      log.error('Tag', 'findAll() target is neither Contact nor Favorite')
+    }
+
+    return tagList
+  }
+
+  /**
+   * Get a new tag
+   *
+   * > Tips:
+   * This function is depending on the Puppet Implementation, see [puppet-compatible-table](https://github.com/Chatie/wechaty/wiki/Puppet#3-puppet-compatible-table)
+   *
+   * @static
+   * @param {string} [tag] the tag name which want to create
+   * @returns {Promise<Tag>}
+   * @example
+   * const bot = new Wechaty()
+   * await bot.Tag.get(tag)
+   */
+  public static async get<T extends typeof Tag> (
+    this: T,
+    tag: string,
+  ): Promise<T['prototype']> {
+    log.verbose('Tag', 'get(%s)', tag)
+    return this.load(tag)
+  }
+
+  /**
+   * Add tag for contact
+   *
+   * > Tips:
+   * This function is depending on the Puppet Implementation, see [puppet-compatible-table](https://github.com/Chatie/wechaty/wiki/Puppet#3-puppet-compatible-table)
+   *
+   * @param {Contact} [to] the contact which need to add tag
+   * @returns {Promise<void>}
+   * @example
+   * await tag.add(contact)
+   */
+  public async add (
+    to: Contact
+  ): Promise<void> {
+    log.verbose('Tag', 'add(%s) for %s', to, this.tag)
+
+    try {
+      await this.puppet.tagAddContact(this.tag, to.id)
+    } catch (e) {
+      log.error('Tag', 'add() exception: %s', e.message)
+      throw new Error(`add error : ${e}`)
+    }
+  }
+
+  /**
+   * Delete this tag
+   *
+   * > Tips:
+   * This function is depending on the Puppet Implementation, see [puppet-compatible-table](https://github.com/Chatie/wechaty/wiki/Puppet#3-puppet-compatible-table)
+   *
+   * @returns {Promise<void>}
+   * @example
+   * await tag.del()
+   */
+  public async del (from: Contact): Promise<void> {
```

Comment:
If we want to delete a tag entirely, how to achieve that  with the current design?

Reply:
We can not do delete tag itself operations in this design.

I think we can just leave tag as they are will be cleaner for the Tag API.

BTW: I think we should rename all `del()` to `remove()` because it's more intuitive?

Reply:
Cleaner Tag API will make Tags in WeChat not clean.

If we don't remove tag which we do not need any more, the tag will make WeChat Tags worse. I think this design will make our user who use tag frequently feel not well.

`remove()` is better for this case.